### ISSUE TITLE
AppHost requires macOS 10.15 while .net core 3.1 claims to support macOS 10.13+

### ISSUE THREAD

**Issue by stakira** [dotnet/runtime#61173]
### Description

1. End user receives a error "You can't use this version of the application <appname> with this version of macos. You have macOS 10.14.6. The application requires macOS 10.15 or later." when installing app from dmg.
2. https://github.com/dotnet/core/blob/main/release-notes/3.1/3.1-supported-os.md claims dotnet core supports 10.13+
3. After looking around for a while. I found that `otool -l <apphost binary>` produces:
```
Load command 10
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.15
      sdk 11.1
   ntools 1
     tool 3
  version 609.8
```
Where minos 10.15 matches the user error.
Then I uninstalled dotnet core sdk 5.0 and installed dotnet core sdk 3.1.414. Build result is the same.

4. It is desirable for the apphost to match the min supported version of corresponding dotnet core SDK. Or provide a way for user to change it.
5. `lipo -archs <apphost>` shows that the binary is not built for arm64, as described in https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary. Building the apphost binary for an old macOS version seems quite possible.

### Configuration
User machine:
macOS 10.14
Build machine:
macOS 11.16
`<TargetFramework>netcoreapp3.1</TargetFramework>`
`dotnet --version --> 3.1.414`
XCode 12.4(12D4e). In project setting of a new C++ project, "Deployment Target" can go as low as 10.9. This is verfied with otool on a build too.

### Regression?
Unknown

### Other information
None

**Comment by vitek-karas** [comment]
@VSadov could you please take a look? Note that this is 3.1 - so it's probably a question how/where do we build the 3.1 servicing patches.

And we should also check that 5.0, 6.0 and 7.0 are built with the right min versions (respective to their supported OS matrix).

**Comment by brunocapu** [comment]
Having quite some trouble with user support also:
>"The application requires macOS 10.15 or later"

@stakira I was able to lower `minos` changing my csproj target to **net5.0**:
```xml
<TargetFramework>net5.0</TargetFramework>
```

With this target I got the self-contained binary with  `minos` **10.12** and without `LC_BUILD_VERSION`:
```
% otool -l <path-to-bin>
...
cmd LC_VERSION_MIN_MACOSX
  cmdsize 16
  version 10.12
      sdk 10.15
```
Versions:
```
% dotnet --version
5.0.402
% xcodebuild -version
Xcode 13.0
Build version 13A233
```

Which __does not make sense__ also. Since the supported min version from **net5.0** is 10.13+
see link [5.0-supported-os](https://github.com/dotnet/core/blob/main/release-notes/5.0/5.0-supported-os.md#macos)

I was not able to test it on older macOS. If some one can, please let us known if it is a fix.

I would like to also ask if is there any workaround for this, any way to force clang `-mmacosx-version-min` flag on build.
Regards

**Comment by VSadov** [comment]
The deployment version for runtime components is specified here:

https://github.com/dotnet/runtime/blob/1a214cab8183794e2ce22fedc700487e579053f6/eng/native/configurecompiler.cmake#L450

I've checked a few binaries, including `apphost` and `singlefilehost` on an ARM64 build from current main and they appear to have correct minos (11.0 for arm64).

I will check what is going on with older releases.

**Comment by VSadov** [comment]
I have checked multiple versions of published 3.1 SDK and it seems something went wrong with min version very recently.

- SDK 3.1.413   and before
use LC_VERSION_MIN_MACOSX  command with
version 10.13
sdk       10.13

- SDK 3.1.414    
uses LC_BUILD_VERSION  command   with
minos    10.15
sdk         11.1

- SDK 3.1.415   (newest)
uses LC_BUILD_VERSION  with
minos     11.0
sdk         11.3


I assume switch to a newer build SDK is intentional, but changes to the minos are not.

**Comment by VSadov** [comment]
I think as a **temporary workaround, if needed**, it may be possible to downgrade build environment to use SDK 3.1.413  (august 3.1.19 .NET Core 3.1 comes with that). 
It will have apphost with minos 10.13 and so will the apps built with it.

I will continue looking into what changed in 414 and 415.

**Comment by VSadov** [comment]
The 5.0 branch requires 10.12   (perhaps a mistake and should require 10.13)

https://github.com/dotnet/runtime/blob/97f0e8a5e4110244224fbc115c8d3dc5c44a360f/eng/native/configurecompiler.cmake#L370

**Comment by VSadov** [comment]
I could not find a place where a min OSX version is passed to the compiler in the old 3.1 repo. 
I think we may have just used a fixed build environment with 10.13 SDK and it was setting 10.13 as the requirement - until the build machines had to be upgraded to new OS/SDK. 

The min version needs to be passed explicitly.

**Comment by VSadov** [comment]
BTW. 6.0 and 7.0 branches seems to be doing the right thing in terms of passing the min version to the build and thus require no fixes. 
It is just the 3.1 servicing branch that need attention, and 5.0 to the lesser extent. (requiring 10.12 is ok, but seems strange when the docs say 10.13)

**Comment by agocke** [comment]
@vsadov I looked and it looks like 3.1 sets this as well https://github.com/dotnet/coreclr/blob/e33a9710e012b1026d3ffb7890def2004003fe86/configurecompiler.cmake#L543

and has for a long time. Looking around, people on Stack Overflow think that this command line option may have stopped working and that the proper way is to set `MACOSX_DEPLOYMENT_TARGET`. Is it possible that's what's broken?

**Comment by VSadov** [comment]
The apphost is not built in coreclr repo in 3.1. The fix would need to be in https://github.com/dotnet/core-setup/tree/release/3.1/src . I do not think it sets anything regarding macos min version and just uses the default, which I guess, changed when build machines were updated.

The fix will be just copying what we have in 5.0 to core-setup, although I think it should use 10.13, because that is what I think we say in the docs and what we set in 6.0 https://github.com/dotnet/runtime/blob/1a214cab8183794e2ce22fedc700487e579053f6/eng/native/configurecompiler.cmake#L450

**Comment by VSadov** [comment]
Actually docs for 3.1 say 10.15+

https://github.com/dotnet/core/blob/main/release-notes/3.1/3.1-supported-os.md

**Comment by VSadov** [comment]
I guess, since we required 10.13 in the past, I will set that.

**Comment by agocke** [comment]
This should be fixed and shipped now in 6.0.3.
